### PR TITLE
change format of the artifact

### PR DESCRIPTION
**Pull request by pasinskim**

### CODE REVIEW

File: Documentation/artifact-format.md [hunk 1]
```diff
@@ -127,7 +127,9 @@ Format: JSON
     {
       ...
     }
-  ]
+  ],
+  "compatibleDevices": ["vexpress-qemu", "beaglebone"],
+  "artifactId": "id"
```
Comment (kacf) [inline Documentation/artifact-format.md]: @maciejmrowiec: I think you said we should standardize on the naming here. Was it `device_types_compatible` and `artifact_id`? Please clarify.

File: Documentation/artifact-format.md [hunk 2]
```diff
@@ -164,37 +171,28 @@ or multiple files listed.  For example:
...
-`type` specified in `header-info`. For `rootfs-image` there is no additional fields 
-required. For instance:
+tar file. The first and the only required entry is the type of the update
+corresponding to the type in `header-info` file.
 
 ```
 {
-  "type": "rootfs-image"
+  "type": "rootfs-image",
+  "compatibleDevices": ["vexpress-qemu", "beaglebone"]
```
Comment (kacf) [inline Documentation/artifact-format.md]: I think we can remove `compatibleDevices` from here then, or? It's redundant.

File: Documentation/artifact-format.md (hunk 1 above)
Comment (maciejmrowiec) [inline Documentation/artifact-format.md]: `device_types_compatible` 👍

Comment (kacf) [inline Documentation/artifact-format.md]: Well, in the lack of anything better, let's call it `artifact_id` then.

File: artifacts.go [hunk 3]
```diff
@@ -71,39 +70,51 @@ func read(aPath string) (parser.Workers, error) {
...
 	ar.Register(&p)
 
-	return ar.Read()
+	_, err = ar.Read()
+	if err != nil {
+		return nil, err
+	}
+	defer ar.Close()
+
+	return ar, nil
```
Comment (kacf) [inline artifacts.go]: This a bit confusing: Are you throwing away the result of the first `Read()`, and then returning the object? I'm unsure if you're doing the right thing here. At least a comment to explain would be nice.

File: artifacts.go [hunk 4]
```diff
@@ -71,39 +70,51 @@ func read(aPath string) (parser.Workers, error) {
...
-		if ri, ok := p.(*parser.RootfsParser); ok {
-			fmt.Printf("  image id: '%s'\n", ri.GetImageID())
-		}
+	fmt.Printf("Mender artifact:\n")
+	fmt.Printf("  Name: %s\n", r.GetArtifactName())
+	fmt.Printf("  Format: %s\n", info.Format)
+	fmt.Printf("  Version: %d\n", info.Version)
+	fmt.Printf("  Compatible devices: '%s'\n", r.GetCompatibleDevices())
 
+	fmt.Printf("\nUpdates:\n")
```
Comment (kacf) [inline artifacts.go]: Should be `Log.Infof()`.

File: artifacts.go [hunk 5]
```diff
@@ -71,39 +70,51 @@ func read(aPath string) (parser.Workers, error) {
...
+	for k, p := range parsers {
+		fmt.Printf("  %s\n", k)
+		fmt.Printf("  Type: '%s'\n", p.GetUpdateType().Type)
 		for _, f := range p.GetUpdateFiles() {
-			fmt.Printf("  update files: name: '%s', size: %d bytes, modified: %s\n",
-				f.Name, f.Size, f.Date)
+			fmt.Printf("  Files:\n")
+			fmt.Printf("    %s\n", f.Name)
+			fmt.Printf("    size: %d\n", f.Size)
+			fmt.Printf("    modified: %s\n", f.Date)
```
Comment (kacf) [inline artifacts.go]: `Log.Infof()`.

File: artifacts.go [hunk 6]
```diff
@@ -149,8 +160,8 @@ func run() error {
 			Usage: "Type of device supported by the update",
 		},
 		cli.StringFlag{
-			Name:  "image-id, i",
-			Usage: "Yocto id of the update image",
+			Name:  "artifact-name, n",
+			Usage: "Id of the artifact",
```
Comment (kacf) [inline artifacts.go]: Perhaps `Name of the artifact`, to be consistent?

File: artifacts_test.go [hunk 7]
```diff
@@ -121,7 +121,7 @@ func TestArtifactsValidate(t *testing.T) {
 func TestArtifactsRead(t *testing.T) {
 	// first create archive, that we will be able to read
 	updateTestDir, _ := ioutil.TempDir("", "update")
-	defer os.RemoveAll(updateTestDir)
+	//defer os.RemoveAll(updateTestDir)
```
Comment (kacf) [inline artifacts_test.go]: Put back? I assume you meant for this to be temporary.

File: metadata/metadata.go [hunk 8]
```diff
@@ -103,7 +110,8 @@ func (hi *HeaderInfo) Write(p []byte) (n int, err error) {
 // TypeInfo provides information of type of individual updates
 // archived in artifacts archive.
 type TypeInfo struct {
-	Type string `json:"type"`
+	Type              string   `json:"type"`
+	CompatibleDevices []string `json:"compatibleDevices"`
```
Comment (kacf) [inline metadata/metadata.go]: Should this be `json:"device_types_compatible"`?

File: metadata/metadata.go [hunk 9]
```diff
@@ -121,89 +129,30 @@ func (ti *TypeInfo) Write(p []byte) (n int, err error) {
...
-}
-
-func (rm *RequiredMetadata) Write(p []byte) (n int, err error) {
-	if err := decode(p, rm); err != nil {
-		return 0, err
-	}
-	return len(p), nil
-}
+// The fields which must exist are update-type dependent. In case of
+// `rootfs-update` image type, `rootfsId` is required.
```
Comment (kacf) [inline metadata/metadata.go]: This is no longer true, is it?

File: artifacts.go (hunk 3 above)
Comment (pasinskim) [inline artifacts.go]: Not really. After reading the whole `ar` structure is filled in with the data. Later on we are using it simply. No further `Read()` is done.

File: artifacts.go (hunk 4 above)
Comment (pasinskim) [inline artifacts.go]: Nope. This is to print artifact details to the console.

File: artifacts.go (hunk 5 above)
Comment (pasinskim) [inline artifacts.go]: We are printing this to the console.

File: artifacts.go (hunk 6 above)
Comment (pasinskim) [inline artifacts.go]: Thanks, fixing!

File: metadata/metadata.go (hunk 8 above)
Comment (pasinskim) [inline metadata/metadata.go]: In fact we are removing this...

File: metadata/metadata.go (hunk 9 above)
Comment (pasinskim) [inline metadata/metadata.go]: Fixing!

File: artifacts.go (hunk 3 above)
Comment (kacf) [inline artifacts.go]: Oh, so `Read()` has side effects? It fills the `ar` structure?

File: artifacts.go (hunk 4 above)
Comment (kacf) [inline artifacts.go]: Ah, from a command line option. Yes, makes sense, I thought it was debug prints.

File: artifacts.go (hunk 5 above)
Comment (kacf) [inline artifacts.go]: Right, makes sense.

**Comment by kacf** [comment]
Additionally I would squash this whole branch before merging. The intermediary review commits are not that useful.

**Pull request merged**